ARM/ARM64 split - part 1

carlocaione opened the pull request:
Some important points:

- To have this passing on CI we need to have https://github.com/zephyrproject-rtos/sdk-ng/pull/325 merged and actively used by CI
- This is going to fail on checkpatch because it is going to recheck all the moved code so we are catching all the past warnings, etc...
- This is going to conflict often
- This PR is only for the AArch64 part. When this is merged I will post another PR taking care of the AArch32 part.
- It was a bit more painful than expected but I also think this is necessary

Closes #32741

carlocaione committed:
arm/arm64: Make ARM64 a standalone architecture

Split ARM and ARM64 architectures.

Details:

- CONFIG_ARM64 is decoupled from CONFIG_ARM (not a subset anymore)
- Arch and include AArch64 files are in a dedicated directory
  (arch/arm64 and include/arch/arm64)
- AArch64 boards and SoC are moved to soc/arm64 and boards/arm64
- AArch64-specific DTS files are moved to dts/arm64
- The A72 support for the bcm_vk/viper board is moved in the
  boards/bcm_vk/viper directory

Signed-off-by: Carlo Caione <ccaione@baylibre.com>

galak reviewed (commented):
Is the `aarch32` removal needed in this PR?  Agree it should be done, but wondering about splitting it out into its own PR?

carlocaione commented:
> Is the `aarch32` removal needed in this PR? Agree it should be done, but wondering about splitting it out into its own PR?

Not strictly needed (this is why I have it in a separate commit). But I can split the PRs for the sake of simplicity.

galak commented:
> Not strictly needed (this is why I have it in a separate commit). But I can split the PRs for the sake of simplicity.

Just thinking to get arm64 split merged to limit this PR as much as possible, and than have followup cleanup PRs would make it quicker to get merged.

carlocaione commented:
> Just thinking to get arm64 split merged to limit this PR as much as possible, and than have followup cleanup PRs would make it quicker to get merged.

No objections. I'll use this PR only for the AArch64 part and once this is merged I can push the remaining part.

carlocaione committed:
arm/arm64: Fix misc and trivials for ARM/ARM64 split

Fix the header guards, comments, github labeler, CODEOWNERS and
MAINTAINERS files.

Signed-off-by: Carlo Caione <ccaione@baylibre.com>

npitre commented:
> No objections. I'll use this PR only for the AArch64 part and once this is merged I can push the remaining part.

+1

galak commented on CODEOWNERS:
```diff
@@ -105,7 +104,7 @@
 /boards/arm/nucleo_f401re/                @idlethread
 /boards/arm/nuvoton_pfm_m487/             @ssekar15
-/boards/arm/qemu_cortex_a53/              @carlocaione
+/boards/arm64/                            @carlocaione
```
We should have this be broken up by board.  Since we only have a small handful its easier to do this now.

carlocaione replied:
OK

carlocaione commented:
Alright, I split the PRs. This PR is now only concerning the AArch64 separation.

galak commented on arch/arm64/Kconfig:
```diff
@@ -0,0 +1,39 @@
+# Copyright (c) 2014-2015 Wind River Systems, Inc.
+# SPDX-License-Identifier: Apache-2.0
+
+menu "ARM Options"
```
"ARM 64 Options"

carlocaione replied:
ACK

galak commented on arch/arm64/Kconfig:
```diff
@@ -0,0 +1,39 @@
+
+config ARM_CUSTOM_INTERRUPT_CONTROLLER
+	bool
+	depends on !CPU_CORTEX_M
```
Is the !CPU_CORTEX_M still needed?

carlocaione replied:
ACK

galak commented on arch/arm64/Kconfig:
```diff
@@ -0,0 +1,39 @@
+	  non-GIC) interrupt controller.
+
+	  A number of Cortex-A and Cortex-R cores (Cortex-A5, Cortex-R4/5, ...)
+	  allow interfacing to a custom external interrupt controller and this
```
Is the help here still relevant for arm64?

carlocaione replied:
yes, it is.

galak commented on arch/arm64/core/Kconfig:
```diff
@@ -28,6 +25,12 @@ config CPU_CORTEX_A72
+	bool
+	help
+	  Include support for the Secure Monitor Call (SMC) and Hypervisor
+	  Call (HVC) instructions on Armv7 and above architectures.
```
armv7?

carlocaione replied:
Yes, this is valid from armv7 up.

galak replied:
I thought arm64 started on armv8, but maybe my memory is wrong.

carlocaione replied:
What I mean is that SMCCC is a standard valid also for armv7, hence the description in the help. But in Zephyr ARMv8 is the only to actually use that.

galak commented on arch/arm64/core/tls.c:
```diff
@@ -0,0 +1,53 @@
+#include <app_memory/app_memdomain.h>
+#include <sys/util.h>
+
+#ifdef CONFIG_CPU_CORTEX_M
```
Remove CONFIG_CPU_CORTEX_M

carlocaione replied:
ACK

galak commented on arch/arm64/include/kernel_arch_data.h:
```diff
@@ -0,0 +1,56 @@
+typedef struct __esf _esf_t;
+typedef struct __basic_sf _basic_sf_t;
+
+#ifdef CONFIG_ARM_MPU
```
is ARM_MPU relevant on arm64?

carlocaione replied:
ACK

galak commented on dts/arm64/broadcom/viper-m7.dtsi:
```diff
@@ -0,0 +1,54 @@
+/*
```
Should this move, its cortex-m7?

carlocaione replied:
ah! good catch. The viper board is shipping an a72 and an m7 so we have some DTS duplicated. I copied the directory but forgot to clean it up.

galak reviewed (changes requested):
Some questions and a few cleanups in a few places.

galak commented:
It feels like the broadcom SoC/board stuff might be useful to have in its own commit since I feel like the split going on there between arm64/arm isn't a straight forward move.

carlocaione commented:
> It feels like the broadcom SoC/board stuff might be useful to have in its own commit since I feel like the split going on there between arm64/arm isn't a straight forward move.

That is going to break bisectability for the Broadcom board though.

galak commented:
> That is going to break bisectability for the Broadcom board though.

At least maybe call out what's happening in the commit message for the broadcom board/soc.

carlocaione commented:
@Abhishek-brcm @sandeepbrcm about the bcm_vk/viper split it would be nice to have your review as well.

carlocaione commented:
@galak there is a way to use zephyrproject-rtos/sdk-ng#325 for CI already?

carlocaione commented:
The compliance test is failing as usual on the typedefs, so that's bogus as usual.
